File: frontend/src/api/pcloud/conformance.test.ts
// The dataset is a directory tree outside this project, so this file reads it
// with Node's filesystem API. The reference pulls in the types for that: the
// app's tsconfig declares only browser-side type packages, which is what keeps
// `node:` imports out of shipped code, and this test is not shipped.
/// <reference types="node" />
import { readdirSync, readFileSync, statSync } from 'node:fs';
import { join } from 'node:path';
import { fileURLToPath } from 'node:url';

import { describe, expect, it } from 'vitest';

import { bookPackagePath, findBookCacheFiles, parseBookCacheFile } from './bookCacheFile';
import type { BookJson, PCloudFileRef, ShelfConfig } from './bookpkg';
import {
  collectBookPackages,
  collectFolders,
  createIgnoreRules,
  createNSFWRules,
  DEFAULT_IGNORED_DIRS,
  findBooksFolder,
  findCoverFile,
  findCurrentSource,
  findShelfConfigFile,
  isBookNSFW,
  isSchemaNewerThanSupported,
  parseBookJson,
  parseShelfConfig,
  toSourceMeta
} from './bookpkg';
import type { PCloudItem } from './types';

/**
 * The pCloud half of the cross-language conformance suite. Its Go twin is
 * `shelf/conformance_test.go`, and both read the same dataset from
 * `shelf/testdata/conformance` — see the README there for the contract.
 *
 * This client re-implements the Go shelf's walk against pCloud's listing API
 * and cannot import any of it, so nothing but this dataset notices when the two
 * drift apart. The unit tests beside it cover this module's own edge cases; what
 * is checked here is only what both implementations must agree on.
 */

/** The expected.json shape this harness understands (schema_version in manifest.json). */
const DATASET_VERSION = 3;

const DATASET_ROOT = fileURLToPath(new URL('../../../../shelf/testdata/conformance/', import.meta.url));

interface CaseManifestEntry {
  name: string;
  covers: string;
}

interface Manifest {
  schema_version: number;
  cases: CaseManifestEntry[];
}

interface ExpectedSource {
  id: string;
  schema_version: number;
  created_at: string;
  comment: string;
  format: string;
  md5_hash: string;
  has_content: boolean;
  line_count: number;
  char_count: number;
  assets: string[];
}

interface ExpectedBook {
  path: string;
  folders: string[];
  id: string;
  title: string;
  format: string;
  authors: string[];
  tags: string[];
  identifiers: Record<string, string>;
  language: string;
  comments: string;
  star: number;
  created_at: string;
  updated_at: string;
  published_at: string;
  cover: string;
  cover_present: boolean;
  schema_version_on_disk: number;
  read_only: boolean;
  nsfw: boolean;
  current_source_field: string;
  current_source: string | null;
  sources: ExpectedSource[];
}

interface ExpectedBookCache {
  name: string;
  usable: boolean;
  book_ids: string[];
}

interface ExpectedReading {
  folders: string[];
  books: ExpectedBook[];
  book_caches: ExpectedBookCache[];
}

function readJsonFile<T>(path: string): T {
  return JSON.parse(readFileSync(path, 'utf8')) as T;
}

/**
 * Turns a case's shelf directory into the recursive listing pCloud would
 * return, plus the file contents that listing only points at.
 *
 * Real ids are irrelevant to what is compared, so they are handed out in walk
 * order; the map back to a real path is what lets the harness "download" a file
 * the way the client does.
 */
function listShelf(shelfDir: string): { root: PCloudItem; contents: Map<number, string> } {
  const contents = new Map<number, string>();
  let nextID = 1;

  const walk = (dir: string, name: string): PCloudItem => {
    const item: PCloudItem = { name, isfolder: true, folderid: nextID++, contents: [] };

    for (const entry of readdirSync(dir, { withFileTypes: true })) {
      const entryPath = join(dir, entry.name);
      if (entry.isDirectory()) {
        item.contents?.push(walk(entryPath, entry.name));
        continue;
      }
      const stat = statSync(entryPath);
      const fileid = nextID++;
      contents.set(fileid, entryPath);
      item.contents?.push({
        name: entry.name,
        isfolder: false,
        fileid,
        size: stat.size,
        modified: stat.mtime.toUTCString()
      });
    }

    return item;
  };

  return { root: walk(shelfDir, 'shelf'), contents };
}

function readListedJson<T>(contents: Map<number, string>, ref: PCloudFileRef): T {
  const path = contents.get(ref.fileid);
  if (!path) {
    throw new Error(`listing has no content for file ${ref.name}`);
  }
  return readJsonFile<T>(path);
}

function readCase(shelfDir: string): ExpectedReading {
  const { root, contents } = listShelf(shelfDir);

  const booksFolder = findBooksFolder(root);
  if (!booksFolder) {
    throw new Error(`${shelfDir} has no books folder`);
  }

  // The shelf's own settings decide which directories are skipped, so they are
  // read before the walk, exactly as the provider does.
  const configRef = findShelfConfigFile(root);
  const config: ShelfConfig = configRef ? parseShelfConfig(readListedJson(contents, configRef)) : {};
  const ignore = createIgnoreRules(config.ignoredDirs ?? DEFAULT_IGNORED_DIRS);
  const isNSFWFolder = createNSFWRules(config.nsfwFolders ?? []);

  const books: ExpectedBook[] = [];
  for (const pkg of collectBookPackages(booksFolder, ignore)) {
    if (!pkg.meta) {
      // A package with no book.json is skipped by the provider that consumes
      // this walk, and the dataset deliberately holds none — see the README.
      continue;
    }
    const meta: BookJson = parseBookJson(readListedJson(contents, pkg.meta));

    books.push({
      path: bookPackagePath(pkg),
      folders: pkg.folders,
      id: meta.id,
      title: meta.title,
      format: meta.format ?? '',
      authors: meta.authors ?? [],
      tags: meta.tags ?? [],
      identifiers: meta.identifiers ?? {},
      language: meta.language ?? '',
      comments: meta.comments ?? '',
      star: meta.star ?? 0,
      created_at: meta.created_at ?? '',
      updated_at: meta.updated_at ?? '',
      published_at: meta.published_at ?? '',
      cover: meta.cover ?? '',
      cover_present: findCoverFile(pkg, meta) !== undefined,
      schema_version_on_disk: meta.schema_version ?? 0,
      read_only: isSchemaNewerThanSupported(meta),
      nsfw: isBookNSFW(isNSFWFolder, pkg.folders, meta),
      current_source_field: meta.current_source ?? '',
      current_source: findCurrentSource(pkg, meta)?.id ?? null,
      sources: pkg.sources.map((source) => {
        const sourceMeta = source.meta
          ? toSourceMeta(readListedJson(contents, source.meta), source.id)
          : undefined;
        return {
          id: source.id,
          schema_version: sourceMeta?.schema_version ?? 0,
          created_at: sourceMeta?.created_at ?? '',
          comment: sourceMeta?.comment ?? '',
          format: sourceMeta?.format ?? '',
          md5_hash: sourceMeta?.md5_hash ?? '',
          has_content: source.content !== undefined,
          line_count: sourceMeta?.line_count ?? 0,
          char_count: sourceMeta?.char_count ?? 0,
          assets: Object.keys(source.assets).sort()
        };
      })
    });
  }
  books.sort((a, b) => (a.path < b.path ? -1 : a.path > b.path ? 1 : 0));

  const bookCaches: ExpectedBookCache[] = findBookCacheFiles(root).map((file) => {
    const cache = parseBookCacheFile(readListedJson(contents, file));
    return {
      name: file.name,
      usable: cache !== null,
      book_ids: cache ? Object.keys(cache.books).sort() : []
    };
  });
  bookCaches.sort((a, b) => (a.name < b.name ? -1 : a.name > b.name ? 1 : 0));

  return { folders: collectFolders(booksFolder, ignore), books, book_caches: bookCaches };
}

const manifest = readJsonFile<Manifest>(join(DATASET_ROOT, 'manifest.json'));

describe('bookpkg conformance dataset', () => {
  it('is the version this harness reads', () => {
    expect(manifest.schema_version).toBe(DATASET_VERSION);
    expect(manifest.cases.length).toBeGreaterThan(0);
  });

  // Without this a case could be added and never run here, which is the one
  // failure mode a conformance suite cannot afford: it would report agreement
  // it never checked.
  it('lists every case directory and nothing else', () => {
    const listed = manifest.cases.map((entry) => entry.name).sort();
    const present = readdirSync(join(DATASET_ROOT, 'cases'), { withFileTypes: true })
      .map((entry) => {
        expect(entry.isDirectory()).toBe(true);
        return entry.name;
      })
      .sort();

    expect(listed).toEqual(present);
    for (const entry of manifest.cases) {
      expect(entry.covers, `case ${entry.name} has no "covers" line`).toBeTruthy();
    }
  });

  it('gives every case a shelf and an expected reading', () => {
    for (const entry of manifest.cases) {
      const children = readdirSync(join(DATASET_ROOT, 'cases', entry.name)).sort();
      expect(children, `case ${entry.name}`).toEqual(['expected.json', 'shelf']);
    }
  });
});

describe.each(manifest.cases)('bookpkg conformance: $name', ({ name }) => {
  it('reads the shelf as the dataset expects', () => {
    const caseDir = join(DATASET_ROOT, 'cases', name);
    const expected = readJsonFile<ExpectedReading>(join(caseDir, 'expected.json'));

    expect(readCase(join(caseDir, 'shelf'))).toEqual(expected);
  });
});

File: shelf/conformance_test.go
package shelf

import (
	"cmp"
	json "encoding/json/v2"
	"errors"
	"os"
	"path"
	"path/filepath"
	"slices"
	"strings"
	"testing"

	"github.com/voilelab/plainshelf/internal/jsonopt"
)

/*
The Go half of the cross-language conformance suite. Its TypeScript twin is
frontend/src/api/pcloud/conformance.test.ts, and both read the same dataset from
testdata/conformance — see the README there for the contract.

The point is not to cover this package (the tests around it do that already) but
to pin the observable reading of a shelf to a file the pCloud client can be held
to as well. That client re-implements this walk against pCloud's listing API and
cannot import any of it, so a change here that is not mirrored there is
otherwise invisible until a phone shows the wrong library.
*/

// conformanceDatasetVersion is the expected.json shape this harness understands,
// matching schema_version in testdata/conformance/manifest.json. Both harnesses
// check it, so a dataset change that outpaces one of them fails loudly instead
// of being read as a behavior difference.
const conformanceDatasetVersion = 1

const conformanceRoot = "testdata/conformance"

type conformanceManifest struct {
	SchemaVersion int `json:"schema_version"`
	Cases         []struct {
		Name   string `json:"name"`
		Covers string `json:"covers"`
	} `json:"cases"`
}

// conformanceReading is what both implementations must report for one case.
type conformanceReading struct {
	Folders     []string               `json:"folders"`
	Books      []conformanceBook      `json:"books"`
	BookCaches []conformanceBookCache `json:"book_caches"`
}

type conformanceBook struct {
	Path                string              `json:"path"`
	Folders              []string            `json:"folders"`
	ID                  string              `json:"id"`
	Title               string              `json:"title"`
	Format              string              `json:"format"`
	Authors             []string            `json:"authors"`
	Tags                []string            `json:"tags"`
	Star                int                 `json:"star"`
	Cover               string              `json:"cover"`
	CoverPresent        bool                `json:"cover_present"`
	SchemaVersionOnDisk int                 `json:"schema_version_on_disk"`
	ReadOnly            bool                `json:"read_only"`
	CurrentSourceField  string              `json:"current_source_field"`
	CurrentSource       *string             `json:"current_source"`
	Sources             []conformanceSource `json:"sources"`
}

type conformanceSource struct {
	ID         string   `json:"id"`
	HasContent bool     `json:"has_content"`
	CharCount  int      `json:"char_count"`
	Assets     []string `json:"assets"`
}

type conformanceBookCache struct {
	Name    string   `json:"name"`
	Usable  bool     `json:"usable"`
	BookIDs []string `json:"book_ids"`
}

func TestBookPackageConformance(t *testing.T) {
	manifest := loadConformanceManifest(t)
	for _, entry := range manifest.Cases {
		t.Run(entry.Name, func(t *testing.T) {
			caseDir := filepath.Join(conformanceRoot, "cases", entry.Name)
			expected := loadConformanceExpectation(t, filepath.Join(caseDir, "expected.json"))
			observed := readConformanceCase(t, filepath.Join(caseDir, "shelf"))
			assertConformanceEqual(t, expected, observed)
		})
	}
}

// TestBookPackageConformanceDatasetIsComplete keeps the manifest and the case
// directories in step. Without it a case could be added and never run, which is
// the one failure mode a conformance suite cannot afford: it would report
// agreement it never checked.
func TestBookPackageConformanceDatasetIsComplete(t *testing.T) {
	manifest := loadConformanceManifest(t)

	listed := make([]string, 0, len(manifest.Cases))
	for _, entry := range manifest.Cases {
		if entry.Covers == "" {
			t.Errorf("case %q has no \"covers\" line; say what it is for", entry.Name)
		}
		listed = append(listed, entry.Name)
	}
	slices.Sort(listed)

	entries, err := os.ReadDir(filepath.Join(conformanceRoot, "cases"))
	if err != nil {
		t.Fatalf("failed to read the conformance cases: %v", err)
	}
	var present []string
	for _, entry := range entries {
		if !entry.IsDir() {
			t.Errorf("unexpected file %q beside the conformance cases", entry.Name())
			continue
		}
		present = append(present, entry.Name())
	}
	slices.Sort(present)

	if !slices.Equal(listed, present) {
		t.Errorf("manifest.json lists %v but the cases directory holds %v", listed, present)
	}

	for _, name := range present {
		children, err := os.ReadDir(filepath.Join(conformanceRoot, "cases", name))
		if err != nil {
			t.Fatalf("failed to read case %q: %v", name, err)
		}
		var got []string
		for _, child := range children {
			got = append(got, child.Name())
		}
		slices.Sort(got)
		if !slices.Equal(got, []string{"expected.json", "shelf"}) {
			t.Errorf("case %q holds %v, want exactly expected.json and shelf/", name, got)
		}
	}
}

func loadConformanceManifest(t *testing.T) conformanceManifest {
	t.Helper()

	var manifest conformanceManifest
	decodeConformanceJSON(t, filepath.Join(conformanceRoot, "manifest.json"), &manifest)

	if manifest.SchemaVersion != conformanceDatasetVersion {
		t.Fatalf("conformance dataset is schema_version %d, this harness reads %d",
			manifest.SchemaVersion, conformanceDatasetVersion)
	}
	if len(manifest.Cases) == 0 {
		t.Fatal("conformance dataset lists no cases")
	}
	return manifest
}

func loadConformanceExpectation(t *testing.T, filePath string) conformanceReading {
	t.Helper()

	var expected conformanceReading
	decodeConformanceJSON(t, filePath, &expected)
	return expected
}

// decodeConformanceJSON refuses unknown fields, so a typo in the dataset is a
// failure rather than a silently unchecked expectation.
func decodeConformanceJSON(t *testing.T, filePath string, target any) {
	t.Helper()

	file, err := os.Open(filePath) //nolint:gosec // fixed test fixture path
	if err != nil {
		t.Fatalf("failed to open %s: %v", filePath, err)
	}
	defer file.Close() //nolint:errcheck // read-only fixture

	if err := json.UnmarshalRead(file, target, json.RejectUnknownMembers(true)); err != nil {
		t.Fatalf("failed to decode %s: %v", filePath, err)
	}
}

// readConformanceCase opens one case's shelf and records what this build reads
// from it.
//
// The shelf is copied to a temp directory first: opening one creates a lock
// file and a scan cache under app/, and a fixture that a test run modifies
// would stop describing what was committed.
func readConformanceCase(t *testing.T, shelfDir string) conformanceReading {
	t.Helper()

	libRoot := filepath.Join(t.TempDir(), "shelf")
	if err := os.CopyFS(libRoot, os.DirFS(shelfDir)); err != nil {
		t.Fatalf("failed to copy the case shelf: %v", err)
	}

	// No BookCacheWriterID: the export would add a cache file of its own to
	// app/ and change what the book_caches expectation describes.
	s := newTestShelf(t, &ShelfConf{LibRoot: libRoot})

	return conformanceReading{
		Folders:     s.collectExportFolders(),
		Books:      readConformanceBooks(t, s),
		BookCaches: readConformanceBookCaches(t, s),
	}
}

// readOnDiskSchemaVersion reads book.json's raw schema_version. The normalized
// BookMeta.SchemaVersion cannot report it: a pre-v1 book and a v1 book both read
// as 1 in memory, and that difference is exactly what the dataset records.
func readOnDiskSchemaVersion(t *testing.T, s *Shelf, bookPath string) int {
	t.Helper()

	file, err := s.dbRoot.Open(path.Join(bookPath, BookMetaFile))
	if err != nil {
		t.Fatalf("open book.json (%s): %v", bookPath, err)
	}
	defer file.Close() //nolint:errcheck // read-only probe

	var meta struct {
		SchemaVersion int `json:"schema_version"`
	}
	if err := json.UnmarshalRead(file, &meta, jsonopt.Read()); err != nil {
		t.Fatalf("decode book.json (%s): %v", bookPath, err)
	}
	return meta.SchemaVersion
}

func readConformanceBooks(t *testing.T, s *Shelf) []conformanceBook {
	t.Helper()

	listings, err := s.ListBooksWithCharCount()
	if err != nil {
		t.Fatalf("ListBooksWithCharCount: %v", err)
	}

	observed := make([]conformanceBook, 0, len(listings))
	for _, listing := range listings {
		book := listing.Book
		meta := book.GetMeta()

		// The normalized SchemaVersion in meta cannot tell a pre-v1 book from a
		// v1 one, and that difference is exactly what the dataset records.
		onDiskSchema := readOnDiskSchemaVersion(t, s, book.PackagePath())

		coverPresent := false
		if meta.Cover != "" {
			if _, _, err := book.OpenCover(); err == nil {
				coverPresent = true
			}
		}

		var currentSource *string
		if source, err := book.ResolveCurrentSource(); err == nil {
			id := source.ID()
			currentSource = &id
		} else if !errors.Is(err, ErrSourceNotFound) {
			t.Fatalf("ResolveCurrentSource(%s): %v", book.PackagePath(), err)
		}

		observed = append(observed, conformanceBook{
			Path:                book.PackagePath(),
			Folders:              orEmpty(listing.Folders),
			ID:                  book.ID(),
			Title:               book.Title(),
			Format:              meta.Format,
			Authors:             orEmpty(meta.Authors),
			Tags:                orEmpty(meta.Tags),
			Star:                meta.Star,
			Cover:               meta.Cover,
			CoverPresent:        coverPresent,
			SchemaVersionOnDisk: onDiskSchema,
			ReadOnly:            errors.Is(book.EnsureWritable(), ErrUnsupportedBookSchemaVersion),
			CurrentSourceField:  book.CurrentSource(),
			CurrentSource:       currentSource,
			Sources:             readConformanceSources(t, s, book),
		})
	}

	slices.SortFunc(observed, func(a, b conformanceBook) int { return cmp.Compare(a.Path, b.Path) })
	return observed
}

func readConformanceSources(t *testing.T, s *Shelf, book *Book) []conformanceSource {
	t.Helper()

	sources, err := book.ListSource()
	if err != nil {
		// A book that never had a source has no sources/ folder at all.
		if errors.Is(err, os.ErrNotExist) {
			return []conformanceSource{}
		}
		t.Fatalf("ListSource(%s): %v", book.PackagePath(), err)
	}

	observed := make([]conformanceSource, 0, len(sources))
	for _, source := range sources {
		hasContent := false
		if content, err := source.Open(); err == nil {
			hasContent = true
			content.Close() //nolint:errcheck // read-only probe
		}

		observed = append(observed, conformanceSource{
			ID:         source.ID(),
			HasContent: hasContent,
			CharCount:  source.GetMeta().CharCount,
			Assets:     readConformanceAssets(t, s, source),
		})
	}
	return observed
}

// readConformanceAssets lists a source's illustrations and checks each one is
// reachable under the relative path the format documents. Nothing records the
// contents of assets/ — the filesystem is the list (see asset.go) — so the
// listing itself is what the pCloud client indexes from its own listing.
func readConformanceAssets(t *testing.T, s *Shelf, source *Source) []string {
	t.Helper()

	assetsPath := path.Join(source.FolderPath(), SourceAssetsFolder)
	entries, err := s.dbRoot.ReadDir(assetsPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return []string{}
		}
		t.Fatalf("ReadDir(%s): %v", assetsPath, err)
	}

	names := make([]string, 0, len(entries))
	for _, entry := range entries {
		if entry.IsDir() {
			continue
		}
		name := entry.Name()
		names = append(names, name)

		assetPath, err := source.AssetPath(name)
		if err != nil {
			t.Errorf("AssetPath(%s): %v", name, err)
			continue
		}
		if want := path.Join(assetsPath, name); assetPath != want {
			t.Errorf("AssetPath(%s) = %q, want %q", name, assetPath, want)
		}
		asset, err := source.OpenAsset(name)
		if err != nil {
			t.Errorf("OpenAsset(%s): %v", name, err)
			continue
		}
		asset.File.Close() //nolint:errcheck // read-only probe
	}

	slices.Sort(names)
	return names
}

func readConformanceBookCaches(t *testing.T, s *Shelf) []conformanceBookCache {
	t.Helper()

	entries, err := s.dbRoot.ReadDir(appFolder)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return []conformanceBookCache{}
		}
		t.Fatalf("ReadDir(%s): %v", appFolder, err)
	}

	observed := make([]conformanceBookCache, 0, len(entries))
	for _, entry := range entries {
		name := entry.Name()
		if entry.IsDir() || !strings.HasPrefix(name, bookCacheFilePrefix) || !strings.HasSuffix(name, bookCacheFileSuffix) {
			continue
		}

		record := conformanceBookCache{Name: name, BookIDs: []string{}}
		if cache, err := s.readBookCacheFile(path.Join(appFolder, name)); err == nil {
			record.Usable = true
			for bookID := range cache.Books {
				record.BookIDs = append(record.BookIDs, bookID)
			}
			slices.Sort(record.BookIDs)
		}
		observed = append(observed, record)
	}

	slices.SortFunc(observed, func(a, b conformanceBookCache) int { return cmp.Compare(a.Name, b.Name) })
	return observed
}

// orEmpty replaces a nil slice with an empty one so the JSON comparison below
// sees [] on both sides: the TypeScript harness has no way to express nil.
func orEmpty(values []string) []string {
	if values == nil {
		return []string{}
	}
	return values
}

func assertConformanceEqual(t *testing.T, expected, observed conformanceReading) {
	t.Helper()

	want := marshalConformance(t, expected)
	got := marshalConformance(t, observed)
	if want == got {
		return
	}

	wantLines := strings.Split(want, "\n")
	gotLines := strings.Split(got, "\n")
	for i := range max(len(wantLines), len(gotLines)) {
		wantLine := ""
		if i < len(wantLines) {
			wantLine = wantLines[i]
		}
		gotLine := ""
		if i < len(gotLines) {
			gotLine = gotLines[i]
		}
		if wantLine != gotLine {
			t.Errorf("shelf reading differs from expected.json at line %d:\n  expected: %s\n  read:     %s",
				i+1, wantLine, gotLine)
			break
		}
	}
	t.Errorf("full reading:\nexpected:\n%s\nread:\n%s", want, got)
}

func marshalConformance(t *testing.T, reading conformanceReading) string {
	t.Helper()

	data, err := json.Marshal(reading, jsonopt.Disk())
	if err != nil {
		t.Fatalf("failed to marshal a conformance reading: %v", err)
	}
	return string(data)
}
